File: server/lib/challongePoints.ts
export function calculateChallongePoints(rank: number | null, total: number | null): number {
  if (!rank || !total) return 0;

  if (total >= 49) {
    if (rank === 1) return 400;
    if (rank === 2) return 280;
    if (rank === 3) return 160;
    if (rank === 4) return 120;
    if (rank >= 5 && rank <= 8) return 90;
    if (rank >= 9 && rank <= 12) return 65;
    if (rank >= 13 && rank <= 16) return 50;
    if (rank >= 17 && rank <= 24) return 40;
    if (rank >= 25 && rank <= 32) return 30;
    if (rank >= 33 && rank <= 48) return 15;
    if (rank >= 49) return 10;
  } else if (total >= 33) {
    if (rank === 1) return 350;
    if (rank === 2) return 240;
    if (rank === 3) return 140;
    if (rank === 4) return 110;
    if (rank >= 5 && rank <= 8) return 80;
    if (rank >= 9 && rank <= 12) return 55;
    if (rank >= 13 && rank <= 16) return 40;
    if (rank >= 17 && rank <= 24) return 30;
    if (rank >= 25 && rank <= 32) return 15;
    if (rank >= 33) return 10;
  } else if (total >= 25) {
    if (rank === 1) return 300;
    if (rank === 2) return 200;
    if (rank === 3) return 120;
    if (rank === 4) return 90;
    if (rank >= 5 && rank <= 8) return 70;
    if (rank >= 9 && rank <= 12) return 45;
    if (rank >= 13 && rank <= 16) return 30;
    if (rank >= 17 && rank <= 24) return 15;
    if (rank >= 25) return 10;
  } else if (total >= 17) {
    if (rank === 1) return 250;
    if (rank === 2) return 160;
    if (rank === 3) return 100;
    if (rank === 4) return 80;
    if (rank >= 5 && rank <= 8) return 60;
    if (rank >= 9 && rank <= 12) return 30;
    if (rank >= 13 && rank <= 16) return 15;
    if (rank >= 17) return 10;
  } else if (total >= 13) {
    if (rank === 1) return 200;
    if (rank === 2) return 120;
    if (rank === 3) return 80;
    if (rank === 4) return 60;
    if (rank >= 5 && rank <= 8) return 30;
    if (rank >= 9 && rank <= 12) return 15;
    if (rank >= 13) return 10;
  } else if (total >= 8) {
    if (rank === 1) return 150;
    if (rank === 2) return 80;
    if (rank === 3) return 60;
    if (rank === 4) return 40;
    if (rank >= 5 && rank <= 8) return 20;
    if (rank >= 9) return 10;
  } else if (total >= 6) {
    if (rank === 1) return 100;
    if (rank === 2) return 70;
    if (rank === 3) return 50;
    if (rank === 4) return 30;
    if (rank >= 5) return 10;
  }

  return 0;
}

File: backend-py/app/lib/challonge_points.py
"""Challonge placement scoring, ported from backend/src/lib/challongePoints.ts.

Points depend on both the finishing rank and how large the bracket was, so a
win in a 50-player event is worth far more than a win in a 6-player one. The
tables are the ones the community agreed on; they are data, not a formula.
"""

# (minimum participant count, [(max rank, points), ...]) ordered from the
# largest bracket down. The first bracket size that fits is the one used.
_TABLES: list[tuple[int, list[tuple[int, int]]]] = [
    (49, [(1, 400), (2, 280), (3, 160), (4, 120), (8, 90), (12, 65), (16, 50), (24, 40), (32, 30), (48, 15), (10**9, 10)]),
    (33, [(1, 350), (2, 240), (3, 140), (4, 110), (8, 80), (12, 55), (16, 40), (24, 30), (32, 15), (10**9, 10)]),
    (25, [(1, 300), (2, 200), (3, 120), (4, 90), (8, 70), (12, 45), (16, 30), (24, 15), (10**9, 10)]),
    (17, [(1, 250), (2, 160), (3, 100), (4, 80), (8, 60), (12, 30), (16, 15), (10**9, 10)]),
    (13, [(1, 200), (2, 120), (3, 80), (4, 60), (8, 30), (12, 15), (10**9, 10)]),
    (8, [(1, 150), (2, 80), (3, 60), (4, 40), (8, 20), (10**9, 10)]),
    (6, [(1, 100), (2, 70), (3, 50), (4, 30), (10**9, 10)]),
]


def calculate_challonge_points(rank: int | None, total: int | None) -> int:
    """Points for finishing `rank` out of `total`. Zero if either is missing."""
    if not rank or not total:
        return 0

    for minimum, table in _TABLES:
        if total >= minimum:
            for max_rank, points in table:
                if rank <= max_rank:
                    return points
            return 0

    # Fewer than six participants scores nothing.
    return 0
